src/plat/pc99/machine/dev/vga/io.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2024, DornerWorks
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ */
+#include <config.h>
+#include <arch/kernel/boot_sys.h>
+#include <arch/model/statedata.h>
+#include <machine/io.h>
+#include <plat/machine/io.h>
+#include <drivers/uart.h>
+
+#define VGA_PHYS 0xb8000 // Text buffer address
+
+//black background, light grey text, no blinking
+#define VGA_COLOR 0x07
+#define CLEAR_SCREEN_CHAR 32
+static uint64_t vga_addr = 0;
+static uint64_t vga_size = 0;
+// assume 80x25 console
+#define MAX_X_POS 80
+#define MAX_Y_POS 25
+
+/* Name kept 'serial_init' since other code expects to call this */
+void serial_init(uint16_t port)
+{
+    set_vga_addr(VGA_PHYS);
+}
+
+void set_vga_addr(uint64_t addr)
+{
+    vga_addr = addr;
+    // multiply by 2 since each slot in the VGA buffer has a char byte and 'style' byte
+    vga_size = MAX_X_POS * MAX_Y_POS * 2;
+}
+
+uint64_t get_vga_addr(void)
+{
+    return vga_addr;
+}
+
+static int verify_vga_offset(uint64_t offset)
+{
+    if (offset > vga_size) {
+        return -1;
+    }
+    return 0;
+}
+
+static void vga_write(uint16_t data, uint64_t offset)
+{
+    if (!verify_vga_offset(offset)) {
+        *(uint16_t *)(vga_addr + offset) = data;
+    }
+}
+
+static uint16_t vga_read(uint64_t offset)
+{
+    if (!verify_vga_offset(offset)) {
+        return *(uint16_t *)(vga_addr + offset);
+    }
+    return 0;
+}
+
+static int current_ypos = MAX_Y_POS, current_xpos = 0;
+
+/***
+ * Slight rework of function 'early_vga_write' from
+ * the linux kernel at arch/x86/kernel/early_printk.c
+ ***/
+static void vga_putchar(char c)
+{
+    int  row, prev_column, column;
+    if (current_ypos >= MAX_Y_POS) { // do we need to scroll
+        /* scroll 1 line up */
```

Comment:
Might be simpler to do a `memmove()` here. Also, I think the names for row and columns are mixed up. Maybe just use x/y for the variables to stick to one set of names.

Reply:
I would prefer not to use `memmove()`. The correspondence to the Linux implementation is still pretty clear here and this is not performance critical.

Reply:
And good catch with the row/column names, you are absolutely right, they are swapped. Columns should be < MAX_X_POS, rows should be < MAX_Y_POS.